File: src/main/java/extractor/features/ProgramFeatures.java
package extractor.features;

import java.util.ArrayList;
import java.util.stream.Collectors;

public class ProgramFeatures {
    private String name;

    private ArrayList<ProgramRelation> features = new ArrayList<>();

    ProgramFeatures(String name) {
        this.name = name;
    }

    @Override
    public String toString() {
        String resultString = name + " " +
                features.stream().map(ProgramRelation::toString).collect(Collectors.joining(" "));
        return resultString;
    }

    void addFeature(Property source, String path, Property target) {
        ProgramRelation newRelation = new ProgramRelation(source, target, path);
        features.add(newRelation);
    }

    boolean isEmpty() {
        return features.isEmpty();
    }

}

File: src/main/kotlin/extractors/features/ProgramFeatures.kt
package extractors.features

import java.util.ArrayList
import java.util.StringJoiner

class ProgramFeatures(name: String) {
    val features = ArrayList<ProgramRelation>()
    val name: String

    init {
        this.name = name
    }

    override fun toString(): String {
        val joiner = StringJoiner(" ")
        for (feature in features) {
            val toString = feature.toString()
            joiner.add(toString)
        }
        return name + " " +
                joiner.toString()
    }

    internal fun addFeature(source: Property, path: String, target: Property) {
        val newRelation = ProgramRelation(source, target, path)
        features.add(newRelation)
    }

}
